lib/elixir/lib/enum.ex

```diff
@@ -2537,16 +2580,60 @@ defmodule Enum do
 
   @compile {:inline, enum_to_string: 1}
 
-  defp enumerate_and_count(enumerable, count) when is_list(enumerable) do
-    {enumerable, length(enumerable) - abs(count)}
+  defp enum_to_string(entry) when is_binary(entry), do: entry
+  defp enum_to_string(entry), do: String.Chars.to_string(entry)
+
+  defp enumerate_count_index(enumerable, index) when is_list(enumerable) do
+    length = length(enumerable)
+    case normalize_index(index, length) do
+      i when is_integer(i) and i < length ->
+        {:ok, enumerable, length, i}
+      _ ->
+        :error
+    end
+  end
+
+  defp enumerate_count_index(enumerable, index) do
+    case get_size_index(enumerable, index) do
+      :error ->
+        :error
+
+      {:error, _module} ->
+        {list, count} = map_reduce(enumerable, 0, fn(x, acc) -> {x, acc + 1} end)
+        case normalize_index(index, count) do
+          i when is_integer(i) and i < count ->
+            {:ok, list, count, i}
+          _ ->
+            :error
+        end
+
+      {:ok, size, normalized_index} ->
+        {:ok, to_list(enumerable), size, normalized_index}
+    end
   end
 
-  defp enumerate_and_count(enumerable, count) do
-    map_reduce(enumerable, -abs(count), fn(x, acc) -> {x, acc + 1} end)
+  # Given an enumerable and an index, it returns:
+  # - `{:ok, size, normalized_index}` where normalized_index is an index within bounds
+  #   and zero or positive.
+  # - `:error` if the given index is out of bound
+  # - `{:error, module}` if Enumerable.count/1 is not implemented.
+  defp get_size_index(enumerable, index) do
+    with  {:ok, size} <- Enumerable.count(enumerable),
+          normalized_index when normalized_index < size <- normalize_index(index, size)
+    do
+      {:ok, size, normalized_index}
+    else
+      :error ->
+        :error
+      {:error, module} ->
+        {:error, module}
+    end
   end
 
-  defp enum_to_string(entry) when is_binary(entry), do: entry
-  defp enum_to_string(entry), do: String.Chars.to_string(entry)
+  defp normalize_index(index, count, default \\ :error) do
```

Comment:
true, we are not using now. But i was earlier. It was useful to normalize finish, which is allowed to be out of bounds, so i was passing default as `count - 1`. Since this function could be used by other functions in Enum, I would be more inclined to leave it for now and see if we will use it in other functions. wdyt?